lib/proxy/client.go

```diff
@@ -0,0 +1,563 @@
+	for {
+		select {
+		case <-c.ctx.Done():
+			return
+		case <-ticker.Chan():
+			c.RLock()
+			c.reporter.resetConnections()
+			for _, conn := range c.conns {
+				switch conn.GetState() {
+				case connectivity.Idle:
+					c.reporter.incConnection(c.config.ID, conn.id, connectivity.Idle.String())
+				case connectivity.Connecting:
+					c.reporter.incConnection(c.config.ID, conn.id, connectivity.Connecting.String())
+				case connectivity.Ready:
+					c.reporter.incConnection(c.config.ID, conn.id, connectivity.Ready.String())
+				case connectivity.TransientFailure:
+					c.reporter.incConnection(c.config.ID, conn.id, connectivity.TransientFailure.String())
+				case connectivity.Shutdown:
+					c.reporter.incConnection(c.config.ID, conn.id, connectivity.Shutdown.String())
+				}
+			}
+			c.RUnlock()
+		}
+	}
+}
+
+// sync runs the peer proxy watcher functionality.
+func (c *Client) sync() {
+	proxyWatcher, err := services.NewProxyWatcher(c.ctx, services.ProxyWatcherConfig{
+		ResourceWatcherConfig: services.ResourceWatcherConfig{
+			Component: teleport.Component(teleport.ComponentProxyPeer),
+			Client:    c.config.AccessPoint,
+			Log:       c.config.Log,
+		},
+	})
+	if err != nil {
+		c.config.Log.Errorf("Error initializing proxy peer watcher: %+v.", err)
+		return
+	}
+	defer proxyWatcher.Close()
+
+	for {
+		select {
+		case <-c.ctx.Done():
+			c.config.Log.Debug("Stopping peer proxy sync.")
+			return
+		case proxies := <-proxyWatcher.ProxiesC:
+			if err := c.updateConnections(proxies); err != nil {
+				c.config.Log.Errorf("Error syncing peer proxies: %+v.", err)
+			}
+		}
+	}
+}
+
+func (c *Client) updateConnections(proxies []types.Server) error {
+	c.RLock()
+	var errs []error
+
+	toDial := make(map[string]types.Server)
+	for _, proxy := range proxies {
+		toDial[proxy.GetName()] = proxy
+	}
+
+	var toDelete []string
+	toKeep := make(map[string]*clientConn)
+	for id, conn := range c.conns {
+		proxy, ok := toDial[id]
+
+		// delete nonexistent connections
+		if !ok {
+			toDelete = append(toDelete, id)
+			continue
+		}
+
+		// peer address changed
+		if conn.addr != proxy.GetPeerAddr() {
+			toDelete = append(toDelete, id)
+			continue
+		}
+
+		toKeep[id] = conn
+	}
+
+	for id, proxy := range toDial {
+		// skips itself
+		if id == c.config.ID {
+			continue
+		}
+
+		// skip existing connections
+		if _, ok := toKeep[id]; ok {
+			continue
+		}
+
+		// establish new connections
+		conn, err := c.connect(id, proxy.GetPeerAddr())
+		if err != nil {
+			c.metrics.reportTunnelError(errorProxyPeerTunnelDial)
+			c.config.Log.Debugf("Error dialing peer proxy %+v at %+v", id, proxy.GetPeerAddr())
+			errs = append(errs, err)
+			continue
+		}
+
+		toKeep[id] = conn
+	}
+	c.RUnlock()
+
+	c.Lock()
+	defer c.Unlock()
+
+	for _, id := range toDelete {
+		if conn, ok := c.conns[id]; ok {
+			go c.shutdownConn(conn)
+		}
+	}
+	c.conns = toKeep
+
+	return trace.NewAggregate(errs...)
+}
+
+// Dial dials a node through a peer proxy.
```

Comment:
nit: forgot to update godoc
```suggestion
// DialNode dials a node through a peer proxy.
```